### PR TITLE
Pandas Assignement 🐼 - Alexandre Zajac

### PR DESCRIPTION
**Pull request by alexZajac**
Here are the pandas related questions as well as the visualization for voteA counts.

### CODE REVIEW

**Comment by tomMoral** [comment]
If you are finished, you can sign up for this RAMP challenge: https://ramp.studio/events/brevet_datacamp2020
This is what we are going to do this afternoon. The goal is to try and solve the challenge.
You need to click on the blue button: `Join the event` and I need to approve your participation so send me a message once you have done this.
You can then read the starting kit and go to the left menu to submit your solutions in the sandbox.

File: data/departments.csv [hunk 1]
```diff
@@ -107,4 +107,4 @@ id,region_code,code,name,slug
 106,COM,986,"Wallis et Futuna","wallis et futuna"
 107,COM,987,"Polynésie française","polynesie francaise"
 108,COM,988,Nouvelle-Calédonie,"nouvelle caledonie"
-109,COM,989,"Île de Clipperton","ile de clipperton"
+109,COM,989,"Île de Clipperton","ile de clipperton"
```
Comment (tomMoral) [inline data/departments.csv:110]: Pls don't modify the `csv` file.

File: data/referendum.csv [hunk 2]
```diff
@@ -36789,4 +36789,4 @@ ZZ;FRANCAIS DE L'ETRANGER;7;Europe centrale;89643;54981;318;17055;17289
 ZZ;FRANCAIS DE L'ETRANGER;8;Europe du Sud, Turquie, Israël;109763;84466;292;9299;15706
 ZZ;FRANCAIS DE L'ETRANGER;9;Afrique Nord-Ouest;98997;59887;321;22116;16673
 ZZ;FRANCAIS DE L'ETRANGER;10;Afrique Centre, Sud et Est;89859;46782;566;17008;25503
-ZZ;FRANCAIS DE L'ETRANGER;11;Europe de l'est, Asie, Océanie;80061;42911;488;13975;22687
+ZZ;FRANCAIS DE L'ETRANGER;11;Europe de l'est, Asie, Océanie;80061;42911;488;13975;22687
```
Comment (tomMoral) [inline data/referendum.csv:36792]: ditto

File: pandas_questions.py [hunk 3]
```diff
@@ -11,14 +11,19 @@
 import pandas as pd
 import geopandas as gpd
 import matplotlib.pyplot as plt
+import os
+
+ABROAD_REGION_CODES = ['01', '02', '03', '04', '06', 'COM']
```
Comment (tomMoral) [inline pandas_questions.py:16]: This is clean! I like this.

File: pandas_questions.py [hunk 4]
```diff
@@ -28,8 +33,21 @@ def merge_regions_and_departments(regions, departments):
...
+    modified_departments = departments.drop(['id', 'slug'], axis=1).rename(
+        columns={'code': 'code_dep', 'name': 'name_dep'}
+    )
+    # merge on region code
+    merged_df = pd.merge(
+        modified_regions,
+        modified_departments,
+        left_on="code_reg",
+        right_on="region_code"
+    )
```
Comment (tomMoral) [inline pandas_questions.py:49]: Rmk: Note that you can do the renaming with the merge using `suffixes`.

File: pandas_questions.py [hunk 5]
```diff
@@ -38,8 +56,19 @@ def merge_referendum_and_areas(referendum, regions_and_departments):
     You can drop the lines relative to DOM-TOM-COM departments, and the
     french living abroad.
     """
-
-    return pd.DataFrame({})
+    # pad the code departments 1 -> 01
+    referendum['Department code'] = referendum['Department code'].apply(
+        lambda x: x.zfill(2)
+    )
```
Comment (tomMoral) [inline pandas_questions.py:62]: You can also call directly `df.str.zfill` instead of doing an apply.

File: pandas_questions.py [hunk 6]
```diff
@@ -61,8 +97,28 @@ def plot_referendum_map(referendum_result_by_regions):
       should display the rate of 'Choice A' over all expressed ballots.
     * Return a gpd.GeoDataFrame with a column 'ratio' containing the results.
     """
-
-    return gpd.GeoDataFrame({})
+    # Load the geographic data with geopandas from `regions.geojson`
+    geo_df = gpd.read_file(
+        os.path.join(DIRNAME, "data", "regions.geojson"), sep=";"
+    )
```
Comment (tomMoral) [inline pandas_questions.py:103]: Note that there is a more modern way to handle path with `pathlib` that I recommend if you do this kind of stuff:

```
from pathlib import Path
DIRNAME = Path('.')
filename = DIR / "data" / "regions.geojson"
```

It also contains all the `exists,mkdir,unlink` that are used from `os.path`.

**Review by tomMoral — CHANGES_REQUESTED** [review]
LGTM except the modification of the CSV, please fix those and I will validate this.

File: data/departments.csv (hunk 1 above)
Comment (alexZajac) [reply data/departments.csv]: Added the newline character on input data.

File: data/referendum.csv (hunk 2 above)
Comment (alexZajac) [reply data/referendum.csv]: Added the newline character on input data.

File: pandas_questions.py (hunk 3 above)
Comment (alexZajac) [reply pandas_questions.py:16]: Thanks! Usually, global variables aren't a great idea in complex projects but here it's convenient enough.

File: pandas_questions.py (hunk 4 above)
Comment (alexZajac) [reply pandas_questions.py:45]: Nice improvement, the merge code is cleaner now.

File: pandas_questions.py (hunk 6 above)
Comment (alexZajac) [reply pandas_questions.py]: Nice! I wasn't aware of this python library and handing the paths more convenient this way!

**Review by tomMoral — APPROVED** [review]
LGTM!

File: pandas_questions.py (hunk 3 above)
Comment (tomMoral) [reply pandas_questions.py:16]: It is fine to use global variable for constants.
It is very explicit like this (upper case) and as you don't modify it, it is okay.

File: pandas_questions.py (hunk 6 above)
Comment (tomMoral) [reply pandas_questions.py]: Yep, it quite recent but it is becoming popular :)

**Pull request closed without merging**